### PR TITLE
Blocksparse for gpuarray

### PR DESCRIPTION
**Pull request by abergeron**
This is a port of blocksparse for the new backend.

### CODE REVIEW

**Comment by abergeron** [comment]
This depends on https://github.com/Theano/libgpuarray/pull/149

File: theano/sandbox/gpuarray/blockgemv.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  } else if (out->ga.typecode == GA_DOUBLE) {
+    err = blas_ops->dgemvBatch(cb_fortran, transA,
+                               PyGpuArray_DIMS(out)[2],
+                               PyGpuArray_DIMS(h)[2], 1,
+                               W_list, offW, lda,
+                               inp_list, offInp, PyGpuArray_STRIDES(h)[2] / gpuarray_get_elsize(h->ga.typecode),
+                               1, out_list, offOut, PyGpuArray_STRIDES(out)[2] / gpuarray_get_elsize(out->ga.typecode),
+                               PyGpuArray_DIMS(out)[1] * PyGpuArray_DIMS(h)[1] * PyGpuArray_DIMS(out)[0], 0);
+  } else {
+    err = GA_DEVSUP_ERROR;
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blockgemv.c]: call hgemvBatch and let the back-end return the corresponding error.

File: theano/sandbox/gpuarray/blockger.c [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                              PyGpuArray_DIMS(x)[0] * PyGpuArray_DIMS(x)[1] * PyGpuArray_DIMS(y)[1], 0);
+  } else if (out->ga.typecode == GA_DOUBLE) {
+    err = blas_ops->dgerBatch(cb_fortran,
+                              PyGpuArray_DIMS(y)[2], PyGpuArray_DIMS(x)[2],
+                              *(double *)PyArray_GETPTR1(alpha, 0),
+                              y_list, offY, str_y, x_list, offX, str_x,
+                              o_list, offOut, str_out,
+                              PyGpuArray_DIMS(x)[0] * PyGpuArray_DIMS(x)[1] * PyGpuArray_DIMS(y)[1], 0);
+  } else {
+    err = GA_DEVSUP_ERROR;
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blockger.c]: idem

File: theano/sandbox/gpuarray/blockger.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  }
+  free(o_list);
+  free(offOut);
+  free(x_list);
+  free(offX);
+  free(y_list);
+  free(offY);
+  if (err != GA_NO_ERROR) {
+    PyErr_SetString(PyExc_RuntimeError, "sgerBatch failed");
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blockger.c]: sgerBatch -> gerBatch

File: theano/sandbox/gpuarray/blocksparse.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    def c_headers(self):
+        return ['<gpuarray/buffer_blas.h>', '<gpuarray/buffer.h>',
+                '<gpuarray_helper.h>']
+
+    def make_node(self, o, W, h, inputIdx, outputIdx):
+        ctx = infer_context_name(o, W, h)
+        o = as_gpuarray_variable(o, ctx)
+        W = as_gpuarray_variable(W, ctx)
+        h = as_gpuarray_variable(h, ctx)
+        assert o.ndim == 3
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blocksparse.py]: call as_tensor_variable on inputIdx and outputIdx for now.

Comment (nouiz) [inline theano/sandbox/gpuarray/blocksparse.py]: I mean, in the futur, maybe we want to allow this to be on the GPU, but for now, we don't want to go there.

File: theano/sandbox/gpuarray/blocksparse.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            return [('INPLACE', '1')]
+        else:
+            return []
+
+    def make_node(self, o, x, y, xIdx, yIdx, alpha=None):
+        ctx = infer_context_name(o, x, y)
+        one = tensor.constant(numpy.asarray(1.0, dtype='float32'))
+        o = as_gpuarray_variable(o, ctx)
+        x = as_gpuarray_variable(x, ctx)
+        y = as_gpuarray_variable(y, ctx)
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blocksparse.py]: idem about xIdx, yIdx

File: theano/sandbox/gpuarray/tests/test_blocksparse.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        self.mode = mode_with_gpu.excluding('constant_folding')
+        self.gemv_op = gpu_sparse_block_gemv
+        self.outer_op = gpu_sparse_block_outer
+        self.gemv_class = GpuSparseBlockGemv
+        self.outer_class = GpuSparseBlockOuter
+
+    # This test is temporarily disabled since we disabled the output_merge
+    # and alpha_merge optimizations for blocksparse due to brokeness.
+    # Re-enable when those are re-added.
+    def Xtest_blocksparse_grad_merge(self):
```
Comment (nouiz) [inline theano/sandbox/gpuarray/tests/test_blocksparse.py]: Was this in the old back-end? Do we have this on the CPU?

Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_blocksparse.py]: It was also disabled in the old backend for the same reason and I think it doesn't work on the CPU again for the same reason.

File: theano/sandbox/gpuarray/blockgemv.c [hunk 7]
```diff
@@ -108,8 +108,16 @@ int APPLY_SPECIFIC(blockgemv)(PyGpuArrayObject *o, PyGpuArrayObject *W,
                                inp_list, offInp, PyGpuArray_STRIDES(h)[2] / gpuarray_get_elsize(h->ga.typecode),
                                1, out_list, offOut, PyGpuArray_STRIDES(out)[2] / gpuarray_get_elsize(out->ga.typecode),
                                PyGpuArray_DIMS(out)[1] * PyGpuArray_DIMS(h)[1] * PyGpuArray_DIMS(out)[0], 0);
+  } else if (out->ga.typecode == GA_HALF) {
+    err = blas_ops->sgemvBatch(cb_fortran, transA,
```
Comment (nouiz) [inline theano/sandbox/gpuarray/blockgemv.c]: should be hgemm I think.

**Comment by nouiz** [comment]
I get this error:

RuntimeError: Pygpu was compiled for a different version of the api, recompile it to avoid problems.


Can you add in the error message what it expected and what it got?

**Comment by abergeron** [comment]
I can, but not in this PR.  The message comes from pygpu, not theano.

**Comment by nouiz** [comment]
Test pass most of the time, so merging. But I got once this error. I suppose the problem also exist in the old back-end. We will see if it was just bad luck or it start appearing in the daily buildbot. If that is the case, we will need to make the test more robust to the range of random numbers:

$ DEVICE=cuda2 theano-nose theano/sandbox/gpuarray/tests/test_blocksparse.py 
WARNING (theano.gof.cmodule): not all the following op(s) implement c_code_cache_version(). This makes them recompiled for each process.[<theano.sandbox.gpuarray.dnn.DnnVersion object at 0x7fe47fb892d0>]
Mapped name None to device cuda: GeForce GTX TITAN X (cuDNN version 5004)
.......E..
======================================================================
ERROR: test_sparseblockgemv_grad_1 (theano.sandbox.gpuarray.tests.test_blocksparse.BlockSparse_Gemv_and_Outer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/bastienf/repos/Theano.Clean/theano/tensor/nnet/tests/test_blocksparse.py", line 235, in test_sparseblockgemv_grad_1
    utt.verify_grad(metaop, [b_val, h_val, W_val], mode=self.mode)
  File "/u/bastienf/repos/Theano.Clean/theano/tests/unittest_tools.py", line 91, in verify_grad
    T.verify_grad(op, pt, n_tests, rng, *args, **kwargs)
  File "/u/bastienf/repos/Theano.Clean/theano/gradient.py", line 1709, in verify_grad
    abs_tol, rel_tol)
GradientError: GradientError: numeric gradient and analytic gradient exceed tolerance:
        At position 0 of argument 2,
            abs. error = 1.107789,  abs. tolerance = 0.010000
            rel. error = 1.000000,  rel. tolerance = 0.010000
Exception args: 
The error happened with the following inputs:, [array([[ 1.17346799]], dtype=float32), array([[[ 0.82418805]]], dtype=float32), array([[[[ 0.47996601]]]], dtype=float32)], 
The value of eps is:, None, 
The out_type is:, None

----------------------------------------------------------------------
Ran 10 tests in 13.695s

FAILED (errors=1)

**Pull request merged**